### PR TITLE
Support user defined config file ID

### PR DESCRIPTION
**Pull request by cyrille-leclerc**
Predictable configuration file identifier is key to use the config-file-provider-plugin in Jenkins workflow and even more in `Jenkinsfile`.

The config-file-provider-plugin currently only supports random identifiers (suffixed by `System.currentTimeMillis()`) such as "`org.jenkinsci.plugins.configfiles.maven.MavenSettingsConfig1445959093236`" and the usage in Jenkins workflow is cumbersome: 

```groovy
wrap([$class: 'ConfigFileBuildWrapper', managedFiles: [[
   fileId: 'org.jenkinsci.plugins.configfiles.maven.MavenSettingsConfig1445959093236', 
   targetLocation: './m2-settings.xml']]]) {

    // some block
}
```

Random/unpredictable identifier are even more cumbersome with `Jenkinsfile` as the `Jenkinsfile` is stored with the source code and thus could be executed on any Jenkins server.

This Pull Requests is inspired by the user defined ids introduced for the credentials. 

User define the id of their choice (screenshot 1), use a default generated id (screenshot 2). If the user defined id is already used, a warning message is displayed (screenshot 3).

If the configuration id is empty or if the type of file to create is missing, an error message is displayed (screenshots 4 & 5)

### Screenshot 1: user defined ID

Configuration File Provider Plugin - User Defined ID

### Screenshot 2: default random ID

Configuration File Provider Plugin -Default random ID

### Screenshot 3: warning for already used ID

Configuration File Provider Plugin - Optional ID Suffix

### Screenshot 4: error configuration ID cannot be empty

Configuration File Provider Plugin - Error Config ID Empty

### Screenshot 4: error select file type

Configuration File Provider Plugin - Error Select File Type

### CODE REVIEW

**Comment by cyrille-leclerc** [comment]
cc @jglick

**Comment by cyrille-leclerc** [comment]
If this PR can be merged, I would like to work on getting the config-file-provider plugin to work at the folder level. To allow people to declare config files at the folder level and not only at the top level.

File: src/main/java/org/jenkinsci/plugins/configfiles/ConfigFilesManagement.java [hunk 1]
```diff
@@ -188,25 +184,40 @@ public void doEditConfig(StaplerRequest req, StaplerResponse rsp, @QueryParamete
...
-        for (ConfigProvider provider : ConfigProvider.all()) {
-            if (provider.getProviderId().equals(providerId)) {
-                req.setAttribute("contentType", provider.getContentType());
-                req.setAttribute("provider", provider);
-                Config config = provider.newConfig();
-                req.setAttribute("config", config);
-                break;
-            }
+        if (providerId == null || providerId.isEmpty()) {
+            // TODO add a warning message: you gorgot to select a provider
```
Comment (cyrille-leclerc) [inline src/main/java/org/jenkinsci/plugins/configfiles/ConfigFilesManagement.java]: @jglick can you recommend me a way to add an error message if the user has not selected a provider?

Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/configfiles/ConfigFilesManagement.java]: > can you recommend me a way to add an error message if the user has not selected a provider?

[This](https://github.com/jenkinsci/jenkins/blob/6641190b8cce326bf9040de4d8182669ff7582d3/core/src/main/resources/lib/hudson/newFromList/form.jelly#L102-L119) is the implementation for the _New Item_ page.

**Comment by cyrille-leclerc** [comment]
@imod Hi, I have iterated on the PR with @jglick. The code is in a good shape, can you review this PR and see if it makes sense for you?

**Comment by imod** [comment]
I'll try to take a look at it this weekend - many thanks!

File: src/main/java/org/jenkinsci/lib/configprovider/AbstractConfigProviderImpl.java [hunk 2]
```diff
@@ -54,8 +61,15 @@ public String getProviderId() {
     }
 
     @Override
-    public boolean isResponsibleFor(String configId) {
-        return configId != null && configId.startsWith(getProviderId());
+    public boolean isResponsibleFor(@NonNull String configId) {
+        if (this.configs.containsKey(configId))
```
Comment (imod) [inline src/main/java/org/jenkinsci/lib/configprovider/AbstractConfigProviderImpl.java]: sorry, but I really don't like empty `if`s...

File: src/main/java/org/jenkinsci/lib/configprovider/model/Config.java [hunk 3]
```diff
@@ -39,6 +44,57 @@ of this software and associated documentation files (the "Software"), to deal
...
 	/**
+	 * Get {@link Config} by id.
+	 *
+	 * @param configId if of the desired {@link Config}
+	 * @return desired {@link Config} or {@code null} if not found
+     */
+	@CheckForNull
+	public static Config getByIdOrNull(@Nullable String configId) {
+		if (configId == null)
+			return null;
```
Comment (imod) [inline src/main/java/org/jenkinsci/lib/configprovider/model/Config.java]: dito ^

**Comment by cyrille-leclerc** [comment]
@domi thanks! do you want me to propose to PR to add brackets or you will add them?

**Comment by cyrille-leclerc** [comment]
@domi thanks, I have just seen https://github.com/jenkinsci/config-file-provider-plugin/commit/72c9cb0ba6c4f1be037bd6e3a7e54e7ec3a42757

**Comment by imod** [comment]
@cyrille-leclerc I already did it and also just released a new version of the plugin

**Comment by imod** [comment]
:)

**Pull request merged**